browser/ui/commander/commander_throttle.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/commander/commander_throttle.h"
+
+#include <string>
+#include <vector>
+
+#include "base/functional/bind.h"
+#include "base/functional/callback_forward.h"
+#include "base/location.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_split.h"
+#include "base/task/post_job.h"
+#include "base/task/single_thread_task_runner.h"
+#include "base/task/task_traits.h"
+#include "base/task/thread_pool.h"
+#include "brave/components/commander/common/commander_url.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/commander/commander.h"
+#include "chrome/browser/ui/commander/commander_backend.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/navigation_throttle.h"
+
+namespace commander {
+CommanderThrottle::CommanderThrottle(content::NavigationHandle* handle)
+    : content::NavigationThrottle(handle) {}
+
+CommanderThrottle::~CommanderThrottle() = default;
+
+content::NavigationThrottle::ThrottleCheckResult
+CommanderThrottle::WillStartRequest() {
+  // Don't handle renderer initiated navigations, we don't want them to be able
+  // to trigger commands.
+  // Note: As an additional protection, commander URLs have an
+  // base::UnguessableToken which needs to be set correctly in order to work
+  // properly.
+  if (navigation_handle()->IsRendererInitiated()) {
```

Comment:
i'm not sure this is sufficient. chromium comments say about `IsRendererInitiated`:

```
 // This method returns false for browser-initiated navigations, including:
  //  * any navigation initiated from the omnibox
  //  * navigations via suggestions in browser UI
  //  * navigations via browser UI: Ctrl-R, refresh/forward/back/home buttons
  //  * any other "explicit" URL navigations, e.g. bookmarks
  ```

could it possibly return false for extension-initiated navigations, ex: via webRequest? possibly also a page could trick a user into bookmarking a malicious command without the user seeing what the command is (because dragging any link to the bookmarks bar turns it into a bookmark). 

i would suggest limiting this so it can ONLY be triggered by omnibox user interaction. for instance an event listener for the user hitting enter in the omnibox or clicking a suggestion.

the unguessable scheme helps but maybe that can be leaked via referrer or something like that. it would probably be better if it weren't in the URL at all and instead sent as a request parameter.

Reply:
:+1: yeah I thought that might be the case. I've updated this to not use a throttle anymore, and instead I've patch `OmniboxEditModel`